include/alpaka/kernel/TaskKernelCpuOmp2Blocks.hpp

```diff
@@ -38,12 +39,694 @@
+        struct ParallelFor<TKernel, omp::Schedule>
+        {
+            //! Run parallel OpenMP loop
+            //!
+            //! \tparam TLoopBody The loop body functor type.
+            //! \tparam TIdx The index type.
+            //!
+            //! \param kernel The kernel instance reference
+            //! \param loopBody The loop body functor instance, takes iteration index as input.
+            //! \param numIterations The number of loop iterations.
+            //! \param schedule The schedule object.
+            template<typename TLoopBody, typename TIdx>
+            ALPAKA_FN_HOST void operator()(
+                TKernel const& kernel,
+                TLoopBody&& loopBody,
+                TIdx const numIterations,
+                omp::Schedule const& schedule)
+            {
+                // Forward to ParallelForImpl that performs dispatch by by chunk size
+                switch(schedule.kind)
+                {
+                case omp::Schedule::NoSchedule:
+                    ParallelForImpl<TKernel, omp::Schedule, omp::Schedule::NoSchedule>{}(
+                        kernel,
+                        std::forward<TLoopBody>(loopBody),
+                        numIterations,
+                        schedule);
+                    break;
+                case omp::Schedule::Static:
+                    ParallelForImpl<TKernel, omp::Schedule, omp::Schedule::Static>{}(
+                        kernel,
+                        std::forward<TLoopBody>(loopBody),
+                        numIterations,
+                        schedule);
+                    break;
+                case omp::Schedule::Dynamic:
+                    ParallelForImpl<TKernel, omp::Schedule, omp::Schedule::Dynamic>{}(
+                        kernel,
+                        std::forward<TLoopBody>(loopBody),
+                        numIterations,
+                        schedule);
+                    break;
+                case omp::Schedule::Guided:
+                    ParallelForImpl<TKernel, omp::Schedule, omp::Schedule::Guided>{}(
+                        kernel,
+                        std::forward<TLoopBody>(loopBody),
+                        numIterations,
+                        schedule);
+                    break;
+#    if defined _OPENMP && _OPENMP >= 200805
+                case omp::Schedule::Auto:
+                    ParallelForImpl<TKernel, omp::Schedule, omp::Schedule::Auto>{}(
+                        kernel,
+                        std::forward<TLoopBody>(loopBody),
+                        numIterations,
+                        schedule);
+                    break;
+#    endif
+                case omp::Schedule::Runtime:
+                    ParallelForImpl<TKernel, omp::Schedule, omp::Schedule::Runtime>{}(
+                        kernel,
+                        std::forward<TLoopBody>(loopBody),
+                        numIterations,
+                        schedule);
+                    break;
```

Comment:
The implementation with all the `ParallelForImpl` specializations feels very repretitive. I think I would inline all the `ParallelForImpl` instances into the `switch` statement and declare the loop induction variables outside the `switch` statement. That gets rid of all the specializations, comments, extra argument lists, documentation etc.

Reply:
What do you mean by `inline`? Do you suggest moving the loop implementation to the switch?

Reply:
If `ParallelForImpl` would contain a static method it is also possible to create an array with function pointers to the implementation and than call 
```
functionPtrArray[omp::Schedule::Runtime](kernel,
                        std::forward<TLoopBody>(loopBody),
                        numIterations,
                        schedule);
``` 

instead of using a switch and repeating the code to call the implementation.

Reply:
Yes. I believe this could dramatically shorten the code.

Reply:
Inlining the ParallelForImpl won't work, as you need them also for the static dispatch. As far as I understand, this switch-case code here will be rarely used - normally you use the static dispatch. Nevertheless I''d not use function pointers, since it makes it more difficult for the compiler to optimize.

Reply:
I don't understands your proposal @bernhardmgruber @psychocoderHPC , could you elaborate?

This part is used only when schedule information is set "run-time" via trait specialization. So wrt optimization I don't think it matters much here. The optimized-for-schedule versions are `ParallelForImpl`  ones, they are fully specialized wrt all supported schedules and will be used directly, without `switch`, in case there is no trait (so either use the member or no schedule)

Reply:
Oh sorry, I did not see the second callsite to `ParallelForImpl`. The code is fine then.